always-hydrate.md

```diff
@@ -0,0 +1,133 @@
+would have had `gvfs prefetch-commits` and `gvfs prefetch-files` as two separate commands, which
+would have then led to a cleaner design for this new addition. I'm open to other suggestions here
+if anyone sees a better pattern for extending the current verbs.
+
+### Proposed UX for always-hydrate commands:
+Adding new entries: 
+
+Add an `--always-hydrate` flag to the existing `--files` and `--folders` variants of `gvfs prefetch`.
+This will communicate that whatever list of folders, files, and wildcards that were passed in to the
+command should be saved and reapplied after every projection change
+
+Syntax:
+
+```
+    gvfs prefetch 
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      [--folders-list <path of file containing newline-delimited paths of folders, no wildcards>] 
+      [--[always-]hydrate]
+```
+
+The only change from today's syntax is the optional `[always-]` on the hydrate flag. 
+`--always-hydrate` implies `--hydrate`.
+
+Passing in the `--always-hydrate` flag has all the same behaviors that `prefetch --hydrate` has today
+and also adds the specified files and folders to the `always-hydrate` file, so that those files will
+continue to be hydrated in the future.
+
+Managing existing entries:
+
+In addition to adding entries to the `always-hydrate` file, users will also want to list out the
+existing entries and remove some or all of the existing entries. To that end, we can support the
+following commands:
+
+```
+    gvfs prefetch
+      [<Repo Root Path>]
+      --list-always-hydrate
+```
+
+The `--list-always-hydrate` flag is incompatible with any of the other flags. It will read out the
+`always-hydrate` file and display the entries to the user.
+
+```
+    gvfs prefetch
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      --remove-always-hydrate
+```
+
+The `--remove-always-hydrate` flag is incompatible with any combination of `--[always-]hydrate` or
+`--list-always-hydrate`.
+
+Any entries that match the specified `--files` and `--folders` arguments will be removed from the
+always-hydrate file. To remove all entries, the user can specify `gvfs prefetch 
+--remove-always-hydrate --files *`.
+
+## Implementation
+### Storing always-hydrate entries
+We will create a new file in the `.gvfs` folder called `always-hydrate`. This file is technically
+human readable and editable, but like all files in the `.gvfs` folder, carries no contract around
+maintaining the same format moving forward. All user interactions with this file should go through
+the `gvfs prefetch` command.
+
+Therefore the specific format of the file is not that interesting. For the sake of easy
```

Comment:
I think we should consider guaranteeing a very simple versioned file format. The obvious con is not being able to break it. But in the pro column, we wouldn't need to provide gestures to interact with it beyond a text editor. This helps clear up the awkward `gvfs prefetch --folders x --remove-always-hydrate`

Then tie everything to the file. To turn it off, delete the file. To remove something, edit the file. You can add something by following the pattern. Onboarding still makes sense with a `gvfs prefetch --folders x --always-hydrate` but the flag is an alternate way to say "create this file for me."

Reply:
That was my initial thought as well. However, I really hesitate to go with this approach. Right now, the entire .gvfs folder is technically opaque to users, and this gives us a lot of freedom as we keep adding more features, making things faster, etc. If we open up this one file, not only do we have to guarantee this file's format moving forward, we also risk people not realizing that other files are off limits, and then we can get locked into current formats. Even worse, some of these files are human readable/editable, but will break your repo's functionality if you modify them incorrectly (e.g. the list of modified paths). So I really don't want to open up that rabbit hole for this feature, which is why I settled on saying that the CLI is the only supported way to interact with this feature.

I suppose we could store this file in a location other than .gvfs, but I don't really know where that location would be. In the clone root? In the .git folder? Somewhere else?

The other alternative is to have a new `gvfs always-hydrate` verb. That has its own set of issues, but it may result in a cleaner syntax.

Thoughts?